src/push-state-tree.js

```diff
@@ -106,49 +156,63 @@ var elementPrototype = typeof HTMLElement !== 'undefined' ? HTMLElement : Elemen
 
 function PushStateTree(options) {
   options = options || {};
-  options[USE_PUSH_STATE] = options[USE_PUSH_STATE] !== false;
 
   // Force the instance to always return a HTMLElement
   if (!(this instanceof elementPrototype)) {
     return PushStateTree.apply(PushStateTree.createElement('pushstatetree-route'), arguments);
   }
 
-  this.VERSION = VERSION;
+  this.eventStack = {
+    leave: [],
+    change: [],
+    enter: [],
+    match: []
+  };
 
-  // Setup options
-  for (var prop in options) {
-    if (options.hasOwnProperty(prop)) {
-      this[prop] = options[prop];
+  //TODO: emcapsulate this
+  for (let property in PushStateTree.prototype) {
+    if (typeof PushStateTree.prototype[property] === 'function') {
+      // function wrapper, without bind the context
+      this[property] = PushStateTree.prototype[property].bind(this);
+      continue;
+    }
+    // Copy properties from prototype to the instance
+    if (typeof this[property] == 'undefined') {
+      this[property] = PushStateTree.prototype[property];
     }
   }
 
+  // Initialize internal history when creating a router instance
+  PushStateTree.initInternalHistory();
+
   // Allow switch between pushState or hash navigation modes, in browser that doesn't support
   // pushState it will always be false. and use hash navigation enforced.
   // use backend non permanent redirect when old browsers are detected in the request.
   if (!PushStateTree.hasPushState) {
     proxyReadOnlyProperty(this, USE_PUSH_STATE, false);
   } else {
-    var usePushState = options[USE_PUSH_STATE];
+    let usePushState = true;
     Object.defineProperty(this, USE_PUSH_STATE, {
       get() {
         return usePushState;
       },
       set(val) {
-        usePushState = val !== false;
+        usePushState = PushStateTree.hasPushState ? val !== false : false;
       }
     });
+    this[USE_PUSH_STATE] = options[USE_PUSH_STATE];
   }
 
   // When enabled beautifyLocation will auto switch between hash to pushState when enabled
+  let beautifyLocation = true && this[USE_PUSH_STATE];
```

Comment:
no, I did this in the old times to gain 1kb in the minification, pre-pre-optimizing.... I think it's time to make the code more legible and a bit bigger. ;) Thanks.